benchmark/run_all_benchmarks.sh

```diff
@@ -35,6 +35,11 @@ if [ ! "${FORMATS}" ]; then
     FORMATS="csr,coo,ell,hybrid,sellp"
 fi
 
+if [ ! "${ELL_IMBALANCE_LIMIT}" ]; then
+    echo "ELL_IMBALANCE_LIMIT    environment variable not set - assuming 100" 1>&2
+    ELL_IMBALANCE_LIMIT=100
```

Comment:
I would prefer give `-1` to run all matrices for ell
Also, we have some stuff to give better performance in these cases than the vendors'

Reply:
If you want to run all matrices, you can use a number like `1000000000` which we will definitely not exceed. IMO it does not make any sense to run ELL benchmarks for such imbalanced matrices. ELL can be hundreds to thousands of times slower than any other matrix format, since almost all entries are padding zeroes then.

Reply:
It's slow but we still need the result if we want a complete benchmark. 
`1000000000` should be enough but we do not really know it's enough or not.

Reply:
The max value I found in ginkgo-data/V100 is 100M (note that this is bounded by the number of columns in the matrix), so this is a safe margin, and setting it to 100 would only exclude 450 matrices. Imagine a matrix with a single dense row. Storing it in ELL would mean storing the matrix as Dense. Attempting to build such a matrix would throw `bad_alloc` in the best case and cause the OOM killer to terminate the benchmark in the worst case, leading to incomplete data. Waiting an additional 2 hours for the benchmark to finish for cases that fit into memory just to have complete numbers we already know to be very slow doesn't seem worth it to me.

Reply:
bad_alloc or OOM killer depends on the server setting, but it requires the admin permission.
If those 450 matrices are all OOM case, it should be fine. It's not worth but it needed if we would like to claim we run all real matrices. we should not kick out the matrices when the format in device is available to handle it.
ELL can also handle some cases well with the max/medium > 100, like half of them are 1, some of is 200.
To be clear, it is useful if we just want a quick and close to full benchmark.

Reply:
To give some background: This was motivated by my recent OpenMP benchmarks. I saw runtimes of 400s for matrix conversion, SpMVs failing due to huge CUDA allocations as well as the OOM killer terminating benchmark processes. All of those should be avoided in real benchmarks. max / median being large means that the matrix consists of at least 50% zeros. For 380 of these matrices, the matrix is even at least 75% empty (max / q3), and that is a very low estimate, since many of those have only a single dense row or power-law distributed row sizes. I wouldn't consider ELL a relevant format for these matrices.

Reply:
Now that I think of it, I guess `.max / .mean` is probably a more suitable value, since it actually measures the ratio non-zero vs. empty storage. 100x storage overhead should be a good exclusion criterion, right?

Reply:
In general, yes. I do not judge the functionality. mainly for what should be the default behavior.
For me, the benchmark only needs to run through all inputs and faces error according to the system.
Using `-1` does not need to depend on the matrix_analysis to skip the selection. (need to manually disable the matrix_analysis though)
For CPU allocation, adjusting `/proc/sys/vm/overcommit_memory` might change the behvaior.
the default behavior, malloc does not guarantee the memory is enough. It starts fills the memory when touch it, so it will be killed by OOM. If all are in CUDA, the benchmark can catch the exception and does not kill the program.

Reply:
I am aware of the default allocation behavior on Linux. The thing is, we have two issues here:
1. OOM killer leads to all matrix formats after `ell` not being benchmarked. Modifying system flags is not an issue for us. Might also be that SLURM gets involved, but I haven't checked that in detail.
2. Jobs that don't hit the OOM limit, but allocate huge amounts of memory and take forever (again, I had runtimes of 400 s for a single repetition of a single conversion, i.e. 1h for a _single_ format conversion operation on a _single_ matrix), and we should not run them by default.

Reply:
Currently, you are skipping the matrix if `.max/.median` is above 100 and hence you don't write anything in the `ell { }` for the matrix, right ? Maybe it makes sense to have another parameter `suitability` and set that to `true` or `false` to record these types of situations ?

Reply:
Yes, more precisely `.max / .mean`, since this measures storage overhead. The difference happens only on the parameter level, we do not modify the JSON input, since it only describes format-agnostic information. The different approaches I considered for this are:
1. have an ELL size limit inside the `spmv` benchmark executable itself. That would require us to determine the row size distribution every time, though.
2. do this check only where we have the necessary data available, namely inside `run_all_benchmark.sh` after we computed the `matrix_statistics`.

Where would you put this `suitability` parameter?

Reply:
for the .max/.mean, if the matrices contains dense row contiguously and less(or zero) contiguously, the ell pack should give the similar performance as classical csr.

Reply:
That is only partially true, since the classical CSR kernel only needs to read the non-zeros that are actually there, while the ELL kernel explicitly reads all the zero values and column indices. So the thread-to-row mapping is similar, but the memory access differs a lot.

Reply:
I am also thinking about what is the difference between limiting ell in these cases and limiting cg in symmetric.
ell is designed as matrix format but good for balanced matrix originally.
cg is designed as krylov solver for symmetric matrices.

Reply:
no, our ell has early stopping, so it only touches the first one zero.

Reply:
I think one of the concerns that Mike has, is the uniformity of a benchmark, in the sense that there should always be a output available for all the different parameters (ex. csr, ell, coo etc.) of an operation (spmv) being benchmarked. If you have a category called `suitability` in the parameter, it would make it easier to just say and record that the matrix format was not suitable for that particular operation. So, it would be inside the matrix format itself. So `ell { "suitability": , }`

Reply:
I feel it makes sense to skip these matrices in the benchmarks. It is true that we would outperform the competitors, but this is really not the target problem where we are interested in outperforming competitors. So I am ok with skipping those automatically (even though this breaks - as Mike points out) the assumption that all benchmarks are processed.

Reply:
I see, sorry about the confusion, the check is rather well-hidden and assumes sorted column indices, it seems? We should note that somewhere. But I still feel like we are getting away from the point here. Classical CSR would also be rather unsuitable for this problem, but it does not have any storage overhead compared to other formats, so there is no potential for crashing due to memory limitations.

Reply:
Yes, if it needs the early stopping.
The performance and suitability needs some knowledge about the benchmark set, the format and maybe implementation. the option is advanced option for me. 
@pratikvn comment is good. some output in the json gives the user what happened in the benchmark. also need to add "completed": false. 
for OOM issue, I prefer using some system call to get free space and ell storage, require 2x ell space in conversion and 1x ell + something space in the spmv.

Reply:
I'll be honest: While I understand the intention behind the `skipped` flag, I am not sure it justifies the effort of either
* adding a new flag `skipped_formats` to the benchmarks together with the associated logic or
* modifying JSON data inside the `run_all_benchmarks.sh` script.
I feel like the absense of the `ell` entry speaks for itself. Of course, we should note this in the BENCHMARKING.md.

And as a general comment, we all have limited time to work on things, and this discussion already took us more time (and probably headspace) than I spent creating the entire PR.

For a bit of history, I had similar issues with the SpGEMM benchmarks, where some matrices took huge amounts of time (literally hours) or crashed after a while, causing subsequent benchmarks to fail, and I had to manually exclude them, but here where we have a nice heuristic that can easily identify unsuitable matrices, excluding which avoids any potential for crashes and huge runtimes, so I would heavily prefer doing that, also since we are very much targeting automation with this script.

Reply:
I'm fine with this behavior but agree with Mike that more is needed to make the information useful while plotting and analyzing a complete benchmark run for a paper. Normally, we use `completed` to check which matrix format went through. `completed: false` means that there was an exception, in practice that is almost always an OOM condition. In the best case, we would give a third value to `completed` to something like `skipped` with your condition. Of course, we cannot because `completed` is a boolean.

AFAIR, missing entries altogether in the JSON file make the GPE and other plotter quite annoying to manage for some type of script.

Reply:
What do you think of the latest solution? Moving the ELL imbalance check into formats.hpp plus storing the exception message for `completed: false`? @yhmtsai @pratikvn @tcojean

Reply:
I think it is good. storing exception will be considered in #171. for now it can just throw the exception and mark `completed: false`. For `skipped_formats`, I prefer `-1` to represent non-skipped to reduce the dependency on matrix_statistics

Reply:
ah, I do not check the latest changes. You seems to store the error already?

Reply:
Yes, I had the same idea :) Just not certain whether `true` or `false` should be the default for `keep_error`.